engines/cge2/metaengine.cpp

```diff
@@ -158,6 +158,10 @@ void CGE2MetaEngine::removeSaveState(const char *target, int slot) const {
 	g_system->getSavefileManager()->removeSavefile(fileName);
 }
 
+Common::Language CGE2Engine::getLanguage() const {
+	return _gameDescription->language;
+}
+
```

Comment:
moving it into the cge2 file throws this error message

```
engines/cge2/cge2.cpp:203:25: error: member access into incomplete type
      'const ADGameDescription'
        return _gameDescription->language;
                               ^
```

Reply:
Indeed Dreamweb implements several `DreamWebEngine` functions in meta engine.cpp.
This seems a bit strange to me, but different developers would organise their code in slightly different ways. In the CGE2 engines it was not done this way, so I think it might be better not to introduce engine functions in the metaengine.cpp file.

Any idea why you are getting this error?
Where the the `ADGameDescription` struct declared?

Reply:
it's in engines/advancedDetector.h line 108
```
/**
 * Data table describing a game variant.
 */
struct ADGameDescription {
	/**
	 * The gameid that identifies the game.
	 *
	 * Mainly used for taking the game description from the @ref PlainGameDescriptor table.
	 */
	const char *gameId;

	/**
	 * Field that is used to distinguish between different variants of a game.
	 *
	 * The content of this field is inserted in the generated description for the config file game entry.
	 * If the @ref kADFlagUseExtraAsHint ADFlag is set, the contents of this field are stored
	 * in the config file, and are used to additionally distinguish between game variants.
	 * Also, if the ADGF_USEEXTRAASTITLE game flag is set, the contents of this field
	 * will be used as a description instead the one extracted from the @ref PlainGameDescriptor table.
	 */
	const char *extra;

	/**
	 * A list of individual file entries used for detection.
	 *
	 * 13 files (last is zero-terminated) is the maximum number of files currently used in ScummVM.
	 * A hardcoded number is used here due to a C++ limitation for defining const arrays.
	 */
	ADGameFileDescription filesDescriptions[14];

	/**
	 * Language of the game variant.
	 */
	Common::Language language;

	/**
	 * Platform of the game variant.
	 */
	Common::Platform platform;

	/**
	 * A bitmask of extra flags.
	 *
	 * The top 16 bits are reserved for generic flags defined in @ref ADGameFlags.
	 * This leaves 16 bits to be used by the client code.
	 */
	uint32 flags;

	/**
	 * Game features that are user-controllable.
	 *
	 * This list reflects which features of the GUI should be turned on or off in order to minimize user confusion.
	 * For instance, there is no point in changing game language in single-language games
	 * or have MIDI controls in a game that only supports digital music.
	 */
	const char *guiOptions;
};
```

Reply:
It is indeed.
So do you understand what the error means and how you can solve it?

Reply:
Ah HA!